### PR TITLE
OS-8543 Want support for OFD locks in lx-brand

### PR DESCRIPTION
**Pull request by smokris**
## Problem
On Debian 12 with backports enabled (or Debian Trixie or Debian Sid) running in SmartOS LX, `systemd-sysusers` fails after updating systemd to v254+, leading to failures in `apt upgrade`.

## Steps to reproduce
- Install SmartOS PI 20240418T000439Z
- `imgadm import c5ba4e59-89b1-4804-b89e-b427923490b9` (latest `debian-12` `lx-dataset`)
- `echo '{"brand":"lx","image_uuid":"c5ba4e59-89b1-4804-b89e-b427923490b9","kernel_version":"4.15","resolvers":["1.1.1.1"],"nics":[{"nic_tag":"admin","ips":["dhcp"]}]}' | vmadm create`
- `zlogin uuidFromAboveCommand`
  ```console
  # echo 'deb http://deb.debian.org/debian bookworm-backports main' > /etc/apt/sources.list.d/bookworm-backports.list
  # apt update
  # apt install systemd/bookworm-backports
  …
  Failed to take /etc/passwd lock: Invalid argument
  …
  ```

## Analysis
In [systemd v253 and earlier](https://github.com/systemd/systemd/blob/44777d7ad6b22ab65a9361b2519bc4b516b02eab/src/shared/lockfile-util.c#L26-L56), systemd's file-locking code would try to obtain an Open File Description (OFD) lock via `fcntl(…F_OFD_SETLKW…)`, and when that failed it would fall back to `flock(…F_SETLKW…)`.  However, [systemd v254 dropped the fallback](https://github.com/systemd/systemd/commit/ed70a34ec0f9ac53646101dd9c544b1f60d17e69), so it now _requires_ support for OFD locks.

SmartOS LX does not support OFD locks, so `systemd-sysusers` v254 fails lock `/etc/passwd`, which in turn causes Debian `apt upgrade` to fail.

## Proposed change
Update SmartOS LX's `fcntl` wrapper to use SmartOS's native support for OFD locks.

I have locally built SmartOS with this change, and confirmed that this change enables systemd v254+'s `systemd-sysusers` to run successfully, and thus enables `apt upgrade` to complete.

Before:
```console
# systemd-sysusers
Failed to take /etc/passwd lock: Invalid argument
# echo $?
1
# strace systemd-sysusers
…
fcntl(3, F_OFD_SETLKW, {l_type=F_WRLCK, l_whence=SEEK_SET, l_start=0, l_len=0}) = -1 EINVAL (Invalid argument)
…
```

After:
```console
# systemd-sysusers
# echo $?
0
# strace systemd-sysusers
…
fcntl(3, F_OFD_SETLKW, {l_type=F_WRLCK, l_whence=SEEK_SET, l_start=0, l_len=0}) = 0
…
```

### CODE REVIEW

**Comment by bahamat** [comment]
Filed internally as [OS-8543](https://smartos.org/bugview/OS-8543).

**Comment by danmcd** [comment]
> This could be one of those cases where we lie and just work, but given the description in glibc, I wonder if we need to keep some brand state about this or something.

Apologies for not looking more deeply.  We actually have some of this already in the base OS (aka OS-2868 upstreamed to -gate as illumos#3252), with some caveats.  Here's text from the illumos bug:

```
As part of this Keith Wesolowski originally did a survey of the various differences
between POSIX locks set through fcntl, lockf, flock, and the newer GNU OFD
(open file descriptor) locks which are trying to tackle and more broadly standardize
the notion of locks being set upon an open file descriptor. Based on this, he
determined that a flock is basically a specialized type of OFD, hence implementation
of similar APIs here.
```

So *maybe* we're Good Enough in the illumos#3252 support?

File: usr/src/uts/common/brand/lx/syscall/lx_fcntl.c [hunk 1]
```diff
@@ -465,16 +475,32 @@ lx_fcntl_lock(int fd, int lx_cmd, void *arg)
 		if ((error = flock_check(vp, &bf, offset, maxoffset)) != 0)
 			break;
 
+		if (cmd == F_OFD_GETLK || cmd == F_OFD_SETLK ||
+		    cmd == F_OFD_SETLKW) {
```
Comment (danmcd) [inline usr/src/uts/common/brand/lx/syscall/lx_fcntl.c:479]: Need the comment as is shown in native fcntl.c support:
```
			/*
			 * TBD OFD-style locking is currently limited to
			 * covering the entire file.
			 */
```

Comment (danmcd) [reply usr/src/uts/common/brand/lx/syscall/lx_fcntl.c:479]: And Linux progs may not like that.  It'll be noteworthy, possibly even DTrace-probe-worthy.

**Review by danmcd — COMMENTED** [review]
Need extra comments and maybe even a DTrace probe in the path where when/start/len error-out with OFD operations.

**Comment by danmcd** [comment]
Additionally @smokris ==> Please inspect $UTS/common/syscall/fcntl.c to make sure the differences in LX's fcntl() vs. native's aren't further apart than Linux apps expect them to be.

**Review by danmcd — COMMENTED** [review]
Also block-comment on top of lx_fcntl_lock() mentions 3 commands.  It's now 6.

File: usr/src/uts/common/brand/lx/syscall/lx_fcntl.c [hunk 2]
```diff
@@ -437,6 +444,9 @@ lx_fcntl_lock(int fd, int lx_cmd, void *arg)
 	case F_GETLK:
 	case F_SETLK:
 	case F_SETLKW:
+	case F_OFD_GETLK:
+	case F_OFD_SETLK:
+	case F_OFD_SETLKW:
```
Comment (danmcd) [inline usr/src/uts/common/brand/lx/syscall/lx_fcntl.c:449]: NOTE:  OFD ops are only allowable on 64-bit or with *large file* support.  For those you'll need to use the 64-bit.  You'll need to have to make checks for F_OFD_* here as it is ONLY supported in 32-bit apps on LARGE FILE ops.  (F_OFD_* is implicitly F_OFD_*64.)  Basically the 32-bit struct size won't work for F_OFD_*.

**Review by danmcd — APPROVED** [review]
I did some looking at both Linux headers and our own native F_OFD_* code.  There's a subtlety in copyin/copyout you're missing.  Otherwise this seems sensible after a first deep glance.  Do you happen to know if there are LInux tests for F_OFD_* anywhere?

Also block-comment on top of lx_fcntl_lock() mentions 3 commands.  It's now 6.

**Comment by danmcd** [comment]
Per an offline conversation with @smokris , I'm going to take over guiding this PR in.  Depending on how cooperative github is, I may need to close this and open a new one.  We'll see.

File: usr/src/uts/common/brand/lx/syscall/lx_fcntl.c (hunk 2 above)
Comment (danmcd) [reply usr/src/uts/common/brand/lx/syscall/lx_fcntl.c:449]: I've determined this is a big mess, and the code-changes underneath here will add more mess.  If I can push to this PR, it'll be clear how and why.

File: usr/src/uts/common/brand/lx/syscall/lx_fcntl.c (hunk 1 above)
Comment (danmcd) [reply usr/src/uts/common/brand/lx/syscall/lx_fcntl.c:479]: Inserted comments and probe into the appropriate spots.

**Review by danmcd — CHANGES_REQUESTED** [review]
Removing +1 to take over this.

File: usr/src/uts/common/brand/lx/syscall/lx_fcntl.c (hunk 1 above)
Comment (danmcd) [reply usr/src/uts/common/brand/lx/syscall/lx_fcntl.c:479]: Now in #471

File: usr/src/uts/common/brand/lx/syscall/lx_fcntl.c (hunk 2 above)
Comment (danmcd) [reply usr/src/uts/common/brand/lx/syscall/lx_fcntl.c:449]: Now in #471

**Comment by danmcd** [comment]
Closing PR in lieu of #471

**Pull request closed without merging**